control-plane-operator/hostedclusterconfigoperator/controllers/resources/azure/azure.go

```diff
@@ -0,0 +1,121 @@
+		azureClientIDs.azureFile = hcp.Spec.Platform.Azure.AzureAuthenticationConfig.ManagedIdentities.DataPlane.FileMSIClientID
+		azureClientIDs.imageRegistry = hcp.Spec.Platform.Azure.AzureAuthenticationConfig.ManagedIdentities.DataPlane.ImageRegistryMSIClientID
+	} else {
+		azureClientIDs.ingress = string(hcp.Spec.Platform.Azure.AzureAuthenticationConfig.WorkloadIdentities.Ingress.ClientID)
+		azureClientIDs.azureDisk = string(hcp.Spec.Platform.Azure.AzureAuthenticationConfig.WorkloadIdentities.Disk.ClientID)
+		azureClientIDs.azureFile = string(hcp.Spec.Platform.Azure.AzureAuthenticationConfig.WorkloadIdentities.File.ClientID)
+		azureClientIDs.imageRegistry = string(hcp.Spec.Platform.Azure.AzureAuthenticationConfig.WorkloadIdentities.ImageRegistry.ClientID)
+	}
+
+	// Skip this step if the user explicitly disabled ingress.
+	if capabilities.IsIngressCapabilityEnabled(hcp.Spec.Capabilities) {
+		ingress := manifests.AzureIngressCloudCredsSecret()
+		if _, err := upsertProvider.CreateOrUpdate(ctx, client, ingress, func() error {
+			data, err := cloneAndSetClientID(secretData, "ingress", azureClientIDs.ingress)
+			if err != nil {
+				return err
+			}
+
+			ingress.Data = data
+			return nil
+		}); err != nil {
+			errs = append(errs, fmt.Errorf("failed to reconcile guest cluster ingress operator secret: %w", err))
+		}
+	}
+
+	// Skip this step if the user explicitly disabled image registry.
+	if capabilities.IsImageRegistryCapabilityEnabled(hcp.Spec.Capabilities) {
+		imageRegistrySecret := manifests.AzureImageRegistryCloudCredsSecret()
+		if _, err := upsertProvider.CreateOrUpdate(ctx, client, imageRegistrySecret, func() error {
+			data, err := cloneAndSetClientID(secretData, "imageRegistry", azureClientIDs.imageRegistry)
+			if err != nil {
+				return err
+			}
+
+			imageRegistrySecret.Data = data
+			return nil
+		}); err != nil {
+			errs = append(errs, fmt.Errorf("failed to reconcile guest cluster image-registry secret: %w", err))
+		}
+	}
+
+	azureDiskCSISecret := manifests.AzureDiskCSICloudCredsSecret()
+	if _, err := upsertProvider.CreateOrUpdate(ctx, client, azureDiskCSISecret, func() error {
+		data, err := cloneAndSetClientID(secretData, "azureDisk", azureClientIDs.azureDisk)
+		if err != nil {
+			return err
+		}
+
+		azureDiskCSISecret.Data = data
+		return nil
+	}); err != nil {
+		errs = append(errs, fmt.Errorf("failed to reconcile guest cluster CSI secret: %w", err))
+	}
+
+	azureFileCSISecret := manifests.AzureFileCSICloudCredsSecret()
+	if _, err := upsertProvider.CreateOrUpdate(ctx, client, azureFileCSISecret, func() error {
+		data, err := cloneAndSetClientID(secretData, "azureFile", azureClientIDs.azureFile)
+		if err != nil {
+			return err
+		}
+
+		azureFileCSISecret.Data = data
+		return nil
+	}); err != nil {
+		errs = append(errs, fmt.Errorf("failed to reconcile csi driver secret: %w", err))
+	}
```

Comment:
NIT: these four blocks basically do the same thing:
- they build a secret
- call `CreateOrUpdate` passing the object, the name, and a mutating func that basically does the same thing

personally I don't feel the repetition adds much readability given that a for loop or an variadic helper function
would be simple, but that's just my personal preference.